Improve ARIA support

cguglielmo opened the pull request.

cguglielmo committed:
aria: remove field description

The description which explains how to use a field is read every time a
field is focused. This is annoying. The aria-details attribute could be
used for this use case. Unfortunately, it is not possible to link an
element that won't be read in browse mode as it is possible with
describedby (at least for NVDA). To reduce the noise, the details
elements could be created on focus and removed on blur. However, when
pressing up in browse mode it will still read the details and not as
a whole text and not even sentence by sentence which is very confusing.

-> Remove description / details completely

394433

cguglielmo committed:
aria: mark invalid fields and consider status for description and error

Fields with an error now get the attribute aria-invalid and
aria-errormessage which is linked to the content of the displayed error
status.

If a status is set, but it doesn't have severity ERROR, the message is
written into aria-description replacing the tooltip text if there is
one. This also happens visually: a status message replaces the tooltip.

394433

cguglielmo committed:
aria: make field status accessible

The field status now gets the proper aria roles and can be accessed
using tab.

Also:
- Support keyboard navigation for menus in tooltip
  Because a field status can open a tooltip with menus, the menus need
  to support keyboard-navigation
- Menu: support selected state
  A context menu uses the css class 'selected' during keyboard
  navigation to highlight the current menu item. This breaks two things:
  - The regular selected state cannot be visualized correctly.
  - If a tooltip has a hierarchical menu, the menu is opened as context
    menu popup. Closing it again with ESC does not focus the parent menu
    item in the tooltip because the selected state is removed by the
    menu item.

Changes:
- The menu item now uses the class 'focused' instead of 'selected' to
  for the keyboard navigation.
- The class 'selected' now properly visualizes the selected state of a
  menu item. The state is now also correctly synchronized between a
  context menu item and the original menu item.
- If the context menu item only toggles a sub menu or a form popup, the
  selected state must not be synchronized (existing behavior). The logic
  in Menu.ts had to be changed slightly to support this case properly.
- The tooltip now better visualizes the different menu item states with
  different severities

394500

cguglielmo committed:
aria: make icon-only actions accessible

Actions with only an icon need to have a tooltip text or an
aria-attribute that describes the purpose of the action, e.g.
aria-label.

-> Update ellipsis menus for menu bars and tab boxes.

394433

cguglielmo committed:
aria: make remove link in filechooser tabbable

394500

cguglielmo committed:
Action: add exec keystroke by default

It should always be possible to execute a tabbable action by using the
default exec keystrokes (space, enter) -> Always register the keystroke.

394500

cguglielmo committed:
aria: add some string literals for params and jsdoc

394433

cguglielmo committed:
aria: improve accessibility for bread crumb bar

The individual items are now tabbable, visualize the focus correctly
and can be activated using keyboard. They now also get the necessary
aria-attributes.

394500

cguglielmo committed:
aria: improve keyboard-navigation for outlines

The outline can now get the focus by using TAB, it is still not
focusable by mouse. When it has the focus, the focused node can be
changed by using the arrow keys the same way as the selection can be
moved. By pressing space, a focused node can be selected.

The tabbable element on tree and table is now $data instead of
$container. This fixes the tab order, e.g. if the menubar is on top of
the data element.

Also: replace getFocusableElement with get$Focusable.
- Makes the naming consistent with get$Scrollable
- Makes it simpler because it always returns a JQuery element and
  rendered checks are not necessary

394500

cguglielmo committed:
Tree: home-keystroke should only select first node not collapse all

It is confusing if HOME collapses all nodes because END just selects the
last node. Also, in hierarchical tables, HOME just selects the first
node -> Adjust HOME keystroke and add new keystroke CTRL-HOME that
collapses all nodes.

Also:
- Some minor code cleanup in keystrokes
- Tests now always set a modifier when triggering key events (false
  instead of undefined) because this is what the browser does and our
  keystrokes expect it that way.
- Removed hints for + and - keystrokes because the arrow keys are the
  preferred way to expand and collapse nodes -> makes it less messy.

cguglielmo committed:
TabBox: improve focus indicator

Currently, only text color is adjusted a little when an item is focused
which is barely visible.

cguglielmo committed:
Action: set preventClickFocus to true by default

Since an action is now tabbable by default, it could get the focus
when clicked. This is not the case for regular buttons or menus, so
actions should behave the same.

394500

cguglielmo committed:
aria: make desktop and table areas tabbable

Various areas are now tabbable:
- Desktop: view buttons, desktop tabs, tool buttons
- Table: controls and header.
- BreadCrumbBar

It works the same way as for menu bars: pressing tab focuses the first
item, the next item can be focused using arrow right key.

The logic from the menubar has been extracted to a coordinator to make
it reusable. The coordinator can handle actions but also plain jquery
items which is necessary for the table header.

Also:
- The filter icons in the table header menu are now tabbable. They were
  converted to actions to get the toggle keystroke for free. Their
  styling has also been improved: they now have a proper hover effect
  and the same size. Flex-box is used to get a proper DOM order which
  is not possible using float: right.

394500

cguglielmo committed:
Table tile mode: fix dom order, enabled state and focus

- If tile mode is active and the user switches to table mode, the
  enabled state was not applied correctly: the rows still looked enabled
  even if the table was disabled.
- The DOM order of the table data, tile header and accordion were
  not correct which has an impact on the tab index.
- In tile mode, the container is focusable, in table mode the data is.
  If data is removed, the focus will be moved outside the table. This
  needs to be prevented so the table keystrokes still work after
  switching.

cguglielmo committed:
aria: add aria attributes to simple tabs

394500

cguglielmo committed:
Simplify field status

- The FieldStatus now sets the css classes on itself and gets a new
  class called field-status. This makes the status-mixing not mandatory
  anymore and can be removed from the owners like FormField etc.
  However, it can still be used if a status needed but FieldStatus.ts
  cannot be used.
- The FormField now always creates a FieldStatus instance. This makes
  it possible to move status relevant code to FieldStatus.
- The error status classes are not passed to date and time popup anymore
  because they are not needed (a popup does not turn red on an error).
- The FieldStatus can now be disabled. This has just been added for the
  sake of completeness and is not needed yet.

394500

cguglielmo committed:
Popup: fix arrow overlay

If the popup has a left padding, the position of the overlay will be
wrong.
-> The overlay needs to start at 0/0 to make the arrow visible.

cguglielmo committed:
Action: make it tabbable by default

An action typically looks like a regular button so it should be
accessible by keyboard as well.

394500

cguglielmo committed:
Switch: make it tabbable by default

394500

cguglielmo committed:
WidgetField: change scrollable default to false

A widget field is supposed to be a wrapper for a widget. If scrollable
is set to true by default, it is not just a wrapper anymore but
changes a behavior. This may be useful for some cases, but normally
it is not desired.

The main problem is that the embedded widget cannot overflow the
 container if it is scrollable. For example if a Switch is embedded,
 the focus indicator will be cut.

394500

cguglielmo committed:
Form: initial focus won't be restored

If a form is opened as view and deactivated without changing the focus
first, the focus will be on the desktop when the form is activated. It
should be in the field that was focused before the form was deactivated.

The problem was, that the focus tracker (onFocusIn) ignored focus events
that happened during rendering, so the initial focus was not tracked.
Now it is tracked even during rendering, but the form has to make sure
it does not render the initial focus when it is re-rendered
-> it now tries to restore the focus itself and only renders the initial
focus if no focus was stored or if restore did not restore anything.
The widget will also call restoreFocus after the form is rendered, but
this should do nothing because the element is already focused.

437591

cguglielmo committed:
aria: improve rules for table, tree and tile grid

- The tabbable element on tree and table is now $data instead of
  $container. This fixes the tab order, e.g. if the menubar is on top
  of the data element.
  Unfortunately, this introduces a new problem: the screen reader uses
  browse mode instead of navigation mode when a tree is focused. It
  works for the tree if the aria roles are moved to the $data as well.
  It does not work for the table, because the header is now outside the
  element with the grid-role. This can be fixed by using aria-owns with
  the caveat that the header row is announced as the last row even if
  an explicit rowindex would be set. It is not perfect but should be
  good enough.
- The table/tree/grid now set aria-rowcount, aria-rowindex, aria-setsize
  and aria-posinset attributes to make the screen reader aware of the
  rows/nodes/tiles that are currently not rendered (virtual scrolling).
  A regular table needs rowcount and rowindex, a hierarchical table
  and listbox need posinset and setsize. Posinset and setsize describe
  the rows of the current level in a hierarchical table (treegrid).
  Furthermore, setsize needs to be set on every row whereas rowcount
  has to be on the container.
- The link of the cells to a header element has been removed to simplify
  it as it is not mandatory and was wrong anyway: it was always linked
  to the first header instead of the header cell of each column.
  However, the boolean column now links the checkbox to the header text.
  This makes Chrome lighthouse happy. And screen reader now reads the
  state correctly even if the checkbox is unchecked.
- The checkbox role of the tree node's checkbox has been removed. It
  would require aria-label (lighthouse), but actually it is not
  necessary because the node itself is checkable and the screen reader
  announces it correctly.
- Toggling the hierarchicalMode of a table now correctly toggles the
  aria role as well.
- aria-activeDescendant of a table is now correctly set even if a row
  is already selected when the table is rendered.
- The tile grid now gets the listbox aria role if it is selectable.
  The grid role does not work because it has no rows. But actually,
  for the user, it can be seen as a horizontal listbox because the
  user can navigate using left and right arrow keys.

394500

cguglielmo committed:
Widget: add property preventInitialFocus

The property was moved from FormField to Widget so it can be used on
any widget, e.g. on actions. Even though it was possible to use the
css class directly, having a property makes it consistent and a bit
easier to use.

To not break existing code, the property is undefined initially which
means the class won't be removed if it is present. In Java, the property
is still on FormField because the data type would have to be changed to
Boolean which breaks existing code.

cguglielmo committed:
jquery-scout: set tabindex to -2 if a scrollable element is not tabbable

cguglielmo committed:
Table/TileGrid: mark focused row/tile

When tabbing on a table with no border and perma-focus  (e.g.
detail-table on desktop), the focus is not visualized. So it is not
obvious, that the table is currently focused.

The detail-table on the desktop now uses a focus-box-shadow similar
to the outline to visualize the focused row.
This also applies for table fields:
- If no rows are selected and tab is used to focus the table data,
  the focused row gets the focus box shadow. This makes more sense
  when tabbing because the previous item is a header item.
- However, if no rows are selected, and the table is clicked or
  programmatically focused, the whole table field gets a focus box
  shadow instead of the focused row, as today (existing behavior).

Some more notes:
- The row focus should only be visible when using the arrow
  keystrokes if it is already visible, which means using the arrow keys
  after clicking into the table should not visualize the row focus.
  This makes it a little cleaner.
- If no row is selected, pressing down does something different if the
  row focus is visible or not. If it is visible, it selects the row
  below the focused row, which is the same behavior as on the tree
  and Windows explorer. If it is not visible, it selects the first row,
  which is the existing behavior.
- Programmatically selecting a row should not automatically focus it
  because it is not possible to determine the correct row when multiple
  rows are selected -> just set it to null. The keystrokes can handle
  this case already.

The same functionality has been implemented for the tile grid.

Also fixed error coloring of table/tree-fields and lookup boxes:
On an error, only the border and shadow of table/tree itself should turn
red, the menu-items or the filter field should stay blue because they
don't have an error.

cguglielmo committed:
Desktop: focus outline content whenever it changes

For example: if a page is selected in the outline, by mouse or keyboard.

Do it only if focus is not already there to not break trackFocus
feature, e.g. if the focus was in a field of a search form, it should
still be in that field when returning to it.
Calling focusManager.validateFocus() is not enough because that would
only request the focus if it is not already in the current focus
context, which is the whole desktop.
So for example if the focus was in the outline, it would stay there when
a page is selected.

369895
394500

cguglielmo committed:
app-links: fix keystroke

The app link handler uses the delegateTarget to read the app link ref
but the AppLinkKeyStroke is often bound to the container and not the
app link directly -> The ref was null for most app links when the
keystroke was used.
However, the click handler is often bound to the app link itself.
This means the target and the delegate target are both candidates for
the app link.

Also ensure keystroke stops propagation, otherwise activating it may
close the form or do a drilldown if the link is on a detail form.

438213

cguglielmo committed:
Widget: add tabbable property

Several widgets already had this property -> moved to Widget to reduce
duplicated code. It also improves the flexibility of widgets that
were already tabbable but did have a dedicated property: it allows
setting tabbable to false from the model. It also allows to add an
auto-logic, e.g. set a table only tabbable if it has data, which can
be disabled by another widget, e.g. proposal chooser.

394500

cguglielmo committed:
Table Spec: fix flip-flopping tests

The row[0] (row number :152) was still hiding even though $.fx.off is
set to true in beforeEach().

See also 250bd09238f14b7af1b35ee05936573f4ad18e1c

cguglielmo committed:
Table/Tree/TileGrid: make it tabbable only if it has data

If there are no visible rows, it should not be in the tab list because
no interaction is possible and focus may not be visible.
For example: a detail table (table of a page), doesn't have a border
like a table field does. So when tabbing into a detail table without
data, the focus would not be visible.
However, it should stay focusable to allow the following case:
When using the SearchOutline, the focus should be in the detail table
after searching. But the rows are inserted async, so the table needs to
accept focus even if it has no rows.

369895
394500

cguglielmo committed:
Action: set preventInitialFocus to true

An action should normally not be the initially focused element.
Since an action is now tabbable by default, it could get the initial
focus. But this is typically not intended, the focus should rather be
in an input element.

The menu items in a menubar are treated slightly different by the focus
manager: they won't get the focus initially unless there is no other
element that gets it. If they had preventInitialFocus set to true, they
would never get the initial focus.

The mode of a mode selector was initially focusable before -> set
preventInitialFocus to false to not change existing behavior.

394500

cguglielmo committed:
SmartField: don't add status if not focused

Otherwise, screen reader would read every status of every smartfield
that has a value because the lookup call is executed during init to
resolve the text of the value.

Also, remove proposal status when the popup closes, otherwise the
screen reader reads obsolete information (e.g. if more than 100 rows
are shown and user enters a text that has no proposals, the screen
reader would still read more than 100 proposals).

cguglielmo committed:
Widget: add property preventClickFocus

The property adds the class `unfocusable` to the focusable element of
the widget. To not break existing code, the property is undefined
initially which means the class won't be removed if it is present.

Compared to using the css class directly, the property has the
following advantages:
- it can be set to true on a base class, e.g. Action where _render is
  often re-implemented in a subclass without having the subclass to
  reset the `unfocusable` class itself
- the property can easily be changed without having to access the
  DOM element or even wait for the widget to be rendered.

cguglielmo committed:
Menu: ensure tooltip is not drawn behind the popup

If clicking the menu opens a popup while a tooltip is open, the popup
would be drawn over the tooltip, which looks bad.
-> Make sure the tooltip is closed when the popup opens and won't be
   shown while a popup is open.

394433

cguglielmo committed:
FocusContext: missing keyboard-navigation with activeElement.tabindex=-1

If active element has tabindex=-1 it is not considered tabbable, only
focusable. So the user can click into it and it gets the focus. When
pressing tab, the next tabbable should be focused and get the class
keyboard-navigation. The focusing will likely work because the browser
takes care of it, but the keyboard-navigation class is missing.

439786

cguglielmo committed:
FilterField: make it tabbable if it contains text

The filter field is opened when the user starts typing. If he leaves it,
the filter field stays open and the user can only enter it again by
typing again which changes the filter. If the table or tree data is not
focusable because it is empty, the filter field cannot be entered again
at all using keyboard -> make it tabbable if it contains text.
Also, the tab order should be: table/tree data, filter field, menu bar.

394500

cguglielmo committed:
Action: use tooltip as aria-label if there is no text

Chrome Lighthouse reports and issue if an action has no label even
if it has a description.
It could be solved by setting a text and textVisible to false, but most
actions that only have an icon set the tooltipText for this purpose.
-> Use tooltipText as aria-label if there is no text

394433

bschwarzent commented on eclipse-scout-core/src/util/events.ts:
```diff
@@ -185,6 +186,34 @@ export const events = {
   },
 
+  /**
+   * @returns the app-link which triggered the given event
```
```suggestion
   * @returns the app-link which triggered the given event, or `null` if the app-link could not be identified
```
Because the method name contains a "$", a caller might assume the result to be `$()` otherwise.

bschwarzent commented on eclipse-scout-core/src/style/mixins.less:
```diff
@@ -158,6 +158,24 @@
+   * Draws an ::after element that has the focus-box-shadow.
+   * An ::after element is necessary to add a little horizontal padding.
+   */
+  .link-with-focus-overlay {
```
This doesn't have anything do with app-links, right?

cguglielmo replied:
it can be used for app-links

bschwarzent commented on eclipse-scout-core/src/desktop/outline/Outline.ts:
```diff
@@ -207,30 +210,36 @@ export class Outline extends Tree implements DisplayParent, OutlineModel {
 
-    this.keyStrokeContext.registerKeyStrokes([
+    let outlineKeyStrokeContext = new OutlineKeyStrokeContext(this);
+    outlineKeyStrokeContext.registerKeyStrokes([
```
Why do we use two separate keystroke contexts and don't just use the OutlineKeyStrokeContext for everything?

cguglielmo replied:
OutlineKeyStrokeContenst has a different scope so these keystrokes can be used from anywhere not only if the outline is focused

bschwarzent commented on eclipse-scout-core/src/desktop/outline/Outline.ts:
```diff
@@ -333,13 +342,13 @@ export class Outline extends Tree implements DisplayParent, OutlineModel {
-    this.$container.setTabbable(false);
+    this.get$Focusable().setTabbable(this.nodesFocusable);
+    this.get$Focusable().unfocusable().addClass('prevent-initial-focus');
   }
```
Why are these two lines in _renderEnabled and not just _render? They don't seem to depend on the `enabled` property.

bschwarzent commented on eclipse-scout-core/src/tree/TreeNode.ts:
```diff
@@ -425,6 +418,9 @@ export class TreeNode implements TreeNodeModel, ObjectWithType, FilterElement {
     if ($node.hasClass('parent-of-selected')) {
       cssClass += ' parent-of-selected';
     }
+    if ($node.hasClass('focused')) {
```
Ok, but why is this method called _preserveCssClasses? It seems to be used to remove all classes expect a few. But what are they?

cguglielmo replied:
This is not a new method, the js doc explains it. If you have a better name, we can change it

bschwarzent commented on eclipse-scout-core/src/widget/Widget.ts:
```diff
@@ -2364,16 +2388,11 @@ export class Widget extends PropertyEventEmitter implements WidgetModel, ObjectW
    *          an optional options object. Shorthand version: If a string is passed instead
    *          of an object, the value is automatically converted to the option {@link ScrollToOptions.align}.
```
Improve description of shorthand? While it is technically still a string, it should now only be one defined by the type ScrollToAlignment.

cguglielmo replied:
any string will be converted not just the ones from the alignment. I think it is fine the way it is

bschwarzent commented on eclipse-scout-core/src/table/TableHeader.ts:
```diff
@@ -116,6 +119,19 @@ export class TableHeader extends Widget implements TableHeaderModel {
+        return item;
+      }
+      return new TabbableItem($headerItem);
+    }).get();
```
This trailing get() is confusing. Consider converting it first to a "normal" array:

`this.findHeaderItems().toArray().map(headerItem => {...});`

bschwarzent commented on eclipse-scout-core/src/table/TableHeader.ts:
```diff
@@ -641,22 +661,30 @@ export class TableHeader extends Widget implements TableHeaderModel {
+    }
+  }
+
+  doHeaderItemAction($headerItem: JQuery, toggleSort?: boolean, multiSort?: boolean) {
```
Is it helpful for this method to be public (without jsdoc)?

bschwarzent commented on eclipse-scout-core/src/keystroke/tabbable/TabbableCoordinator.ts:
```diff
@@ -0,0 +1,345 @@
+  scout, scrollbars, SomeRequired, Widget
+} from '../..';
+
+export class TabbableCoordinator extends PropertyEventEmitter implements TabbableCoordinatorModel {
```
Add a jsdoc description of this class

bschwarzent commented on eclipse-scout-core/src/tabbox/SimpleTabOverflowMenu.ts:
```diff
@@ -0,0 +1,74 @@
+ */
+import {ContextMenuPopup, Menu, MenuModel, Popup, scout, SimpleTabArea} from "..";
+
+/*
```
Remove this duplicate license header

bschwarzent commented on eclipse-scout-core/src/tabbox/SimpleTabOverflowMenu.ts:
```diff
@@ -0,0 +1,74 @@
+ */
+export class SimpleTabOverflowMenu extends Menu implements SimpleTabOverflowMenuModel {
+  declare parent: SimpleTabArea;
+  overflowTabsIndizes: number[];
```
"indizes" is not an English word, also get rid of the inner "s" -> overflowTabIndexes

(The same in SimpleTabAreaLayout.ts)

bschwarzent commented on eclipse-scout-core/src/menu/Menu.ts:
```diff
@@ -385,8 +390,8 @@ export class Menu extends Action implements MenuModel {
-  isTabTarget(): boolean {
-    return this.enabledComputed && this.visible && !this.overflown && (this.isButton() || !this.separator)
+  override isTabTarget(): boolean {
+    return super.isTabTarget() && (this.isButton() || !this.separator) // TODO CGU why check for this.isButton? remove it
```
Resolve TODO

bschwarzent commented on eclipse-scout-core/src/menu/menubar/MenuBar.ts:
```diff
@@ -450,6 +417,7 @@ export class MenuBar extends Widget implements MenuBarModel {
   }
 
   protected _onMenuItemFocus(event: Event<Menu>) {
+    // TODO CGU remove listener
```
Resolve TODO

bschwarzent commented on eclipse-scout-core/src/keystroke/tabbable/FocusPreviousTabTargetKeyStroke.ts:
```diff
@@ -0,0 +1,43 @@
+    let lastValidItem: TabbableItem;
+
+    for (const item of tabbableItems) {
+      if ($focusedItem[0] === item.$container?.[0]) {
```
Could be shortened to `$focusedItem.is(item.$container)`

bschwarzent commented on eclipse-scout-core/src/focus/FocusManager.ts:
```diff
@@ -16,19 +16,20 @@ export interface FocusManagerOptions {
+export interface RequestFocusOptions extends FocusContextFocusOptions {
   /**
-   * prevents focusing if not ready
+   * Prevents focusing if not ready.
```
"Default is false."

bschwarzent commented on eclipse-scout-core/src/desktop/viewbutton/ViewButtonBox.ts:
```diff
@@ -1,13 +1,17 @@
  */
-import {Desktop, Event, EventHandler, HtmlComponent, InitModelOf, ObjectOrChildModel, OutlineViewButton, PropertyChangeEvent, scout, ViewButton, ViewButtonBoxEventMap, ViewButtonBoxModel, ViewMenuTab, Widget, widgets} from '../../index';
+import {
+  aria,
```
Apply formatter

bschwarzent commented on eclipse-scout-core/src/desktop/viewbutton/ViewButtonBox.ts:
```diff
@@ -164,6 +182,12 @@ export class ViewButtonBox extends Widget implements ViewButtonBoxModel {
+
+    // TabbableCoordinator would reset it to initial item if a view button is selected,
+    // but the button of the view menu is still selected and therefore not a tab target yet, so the focus would be on the dropdown -> reset again
+    // this.tabbableCoordinator.resetCurrentItem();
```
Should this be removed?

bschwarzent commented on eclipse-scout-core/src/desktop/viewbutton/ViewButton.ts:
```diff
@@ -99,6 +95,10 @@ export class ViewButton extends Action implements ViewButtonModel {
     this.selectedAsMenu = selectedAsMenu;
   }
 
+  override isTabTarget(): boolean {
```
Remove?

bschwarzent commented on eclipse-scout-core/src/widget/Widget.ts:
```diff
@@ -2111,42 +2148,44 @@ export class Widget extends PropertyEventEmitter implements WidgetModel, ObjectW
+      return this.session.focusManager.requestFocus(this._$lastFocusedElement);
+    }
+    if (this._lastFocusedWidget) {
+      let focused = false;
```
Why not directly assign value?

`let focused = this._lastFocusedWidget.focus();`

bschwarzent commented on eclipse-scout-core/src/aria/aria.ts:
```diff
@@ -167,6 +162,19 @@ export const aria = {
+  /**
+   * @returns the value of the id attribute. If the target doesn't have an id, a new one will be created and assigned.
+   */
+  ensureTargetHasId($target: JQuery<Element>) {
```
- Declare return type
- Is the name "target" relevant? Maybe simplify to `ensureId($element)`

bschwarzent commented on eclipse-scout-core/src/tree/Tree.ts:
```diff
@@ -3039,6 +3090,7 @@ export class Tree extends Widget implements TreeModel, Filterable<TreeNode> {
     }
     result.newlyShown.forEach(node => this._addToVisibleFlatList(node, this.filterAnimated));
     result.newlyHidden.forEach(node => this._removeFromFlatList(node, this.filterAnimated));
+    // this._updateAriaNodeIndices();
```
Remove?

bschwarzent commented on eclipse-scout-core/src/jquery/jquery-scout.js:
```diff
@@ -710,7 +714,7 @@ $.fn._updateTabIndexForReadOnly = function() {
   const readonly = this.attr('readonly');
   // Otherwise make it focusable but not tabbable, because readonly elements are tabbable by default, but we don't want them to be
-  return this.attr('tabindex', readonly ? -1 : null);
+  return this.attr('tabindex', readonly ? -1 : tabindexIfNotTabbable(this));
```
Use consistent argument types -> `'-1'` instead of `-1`
(also in setTabbable)

bschwarzent commented on eclipse-scout-core/src/widget/Widget.ts:
```diff
@@ -933,11 +940,11 @@ export class Widget extends PropertyEventEmitter implements WidgetModel, ObjectW
-      return;
-    }
-    this.$container.setEnabled(this.enabledComputed);
+    this.$container?.setEnabled(this.enabledComputed);
```
Can this.$container really be null in a render-method?

cguglielmo replied:
Not every widget has a $container

fschinkel commented on eclipse-scout-core/src/aria/aria.ts:
```diff
@@ -554,11 +570,27 @@ export const aria = {
+   * @param $elem element to add/remove the attribute. If null, nothing is changed.
+   * @param value value of the attribute to set. If null, attribute is removed.
+   */
+  current($elem: JQuery<Element>, value: 'page' | 'step' | 'location' | 'date' | 'time' | 'true' | 'false' | boolean) {
```
Maybe add a type for this?

fschinkel commented on eclipse-scout-core/src/aria/aria.ts:
```diff
@@ -418,11 +431,11 @@ export const aria = {
    */
-  hasPopup($elem: JQuery<Element>, type: string) {
+  hasPopup($elem: JQuery<Element>, type: 'menu' | 'listbox' | 'tree' | 'grid' | 'dialog' | 'true' | 'false' | boolean) {
     if (!$elem) {
```
Maybe add a type for this?

fschinkel commented on eclipse-scout-core/src/aria/aria.ts:
```diff
@@ -470,7 +483,7 @@ export const aria = {
    */
-  live($elem: JQuery<Element>, value: string) {
+  live($elem: JQuery<Element>, value: 'assertive' | 'polite' | 'off') {
     if (!$elem) {
```
Maybe add a type for this?

fschinkel commented on eclipse-scout-core/src/tooltip/Tooltip.ts:
```diff
@@ -1,13 +1,17 @@
-import {arrays, Desktop, Form, graphics, InitModelOf, keys, Menu, ObjectOrChildModel, Point, Rectangle, scout, scrollbars, Status, StatusSeverity, strings, TooltipEventMap, TooltipModel, Widget, WidgetPopup} from '../index';
+import {
+  aria,
+  arrays, Desktop, Form, graphics, InitModelOf, keys, KeyStrokeContext, Menu, menuNavigationKeyStrokes, ObjectOrChildModel, Point, Rectangle, scout, scrollbars, Status, StatusSeverity, strings, TooltipEventMap, TooltipModel, Widget,
```
Format

fschinkel commented on eclipse-scout-core/test/form/fields/datefield/DateFieldSpec.ts:
```diff
@@ -1553,6 +1553,111 @@ describe('DateField', () => {
+        label: 'label',
+        errorStatus: {
+          objectType: Status,
+          severity: Status.Severity.OK,
```
not necessary

fschinkel commented on eclipse-scout-core/test/form/fields/datefield/DateFieldSpec.ts:
```diff
@@ -1553,6 +1553,111 @@ describe('DateField', () => {
+        label: 'label',
+        errorStatus: {
+          objectType: Status,
+          severity: Status.Severity.ERROR,
```
not necessary

fschinkel commented on eclipse-scout-core/src/action/Action.ts:
```diff
@@ -8,7 +8,9 @@
 import {
-  AbstractLayout, ActionEventMap, ActionKeyStroke, ActionModel, Alignment, aria, Device, DoubleClickSupport, EnumObject, HtmlComponent, Icon, InitModelOf, KeyStrokeContext, LoadingSupport, NullLayout, scout, TooltipPosition, tooltips,
+  AbstractLayout, ActionEventMap, ActionExecKeyStroke, ActionKeyStroke, ActionModel, Alignment, aria, Device, DoubleClickSupport, EnumObject, HtmlComponent, Icon, InitModelOf, KeyStrokeContext, LoadingSupport, MenuExecKeyStroke, NullLayout,
+  scout,
```
Format

fschinkel commented on eclipse-scout-core/test/widget/WidgetSpec.ts:
```diff
@@ -1970,6 +1970,77 @@ describe('Widget', () => {
+  });
+
+
   describe('prepareModel', () => {
```
format

fschinkel commented on eclipse-scout-core/test/form/FormSpec.ts:
```diff
@@ -9,7 +9,8 @@
 import {FormSpecHelper, OutlineSpecHelper, SpecForm} from '../../src/testing/index';
 import {
-  App, CancelMenu, CloseMenu, Dimension, fields, FileChooser, Form, FormFieldMenu, FormModel, InitModelOf, MessageBox, NotificationBadgeStatus, NullWidget, NumberField, ObjectFactory, OkMenu, Popup, PopupBlockerHandler, Rectangle,
+  App, CancelMenu, CloseMenu, Dimension, fields, FileChooser, Form, FormFieldMenu, FormModel, InitModelOf, MessageBox, NotificationBadgeStatus, NullWidget, NumberField, ObjectFactory, OkMenu, Outline, Page, Popup, PopupBlockerHandler,
```
Format

cguglielmo committed:
Action: make it tabbable by default

An action typically looks like a regular button so it should be
accessible by keyboard as well.

394500

cguglielmo committed:
aria: make table footer tabbable

- The info items (load, filter, selection, status) are now tabbable.
- Info items are now actions so they can be activated using keyboard
  without requiring keyboard handlers.
- Adjusted CSS to improve the visualization of the focus border.
  The new margin requires a CSS transition to get a smooth animation.
  Doing it with JavaScript as the current width animation would be
  complicated because the TableFooter takes over and would have to
  animate the margin as well -> CSS transition makes it easier and is
  the preferred way anyway.
- Removed attention animation because it did not work. It could be fixed
  by adding display: inline-block; but it does not really draw attention
  because it is too inconspicuous.
- Made table status tooltips focusable so the tooltip menus can be
  accessed using keyboard.

394500
